File: src/com/lfd/jdsaa/dlink/SingleLink.java
package com.lfd.jdsaa.dlink;

import java.util.Iterator;

public class SingleLink<T> implements Iterable<T>,Iterator<T>{

    private SNode<T> root;
    private SNode<T> itNode;
    private int size;

    public SingleLink<T> append(T data) {
        SNode<T> node = new SNode<>(data);
        if (root == null) {
            root = node;
        } else {
            SNode<T> cNode = root;
            while (cNode.next != null) {
                cNode = cNode.next;
            }
            cNode.next = node;
        }
        this.size ++;
        return this;
    }

    public int size() {
        return size;
    }

    public boolean hasNext(){
        if(size == 0) return false;
        return this.itNode != null;
    }

    @Override
    public T next() {
        if(this.itNode == null){
            throw new RuntimeException("out of range");
        }
        T value = this.itNode.value();
        this.itNode = this.itNode.next;
        return value;
    }


    @Override
    public Iterator<T> iterator() {
        this.itNode = this.root;
        return this;
    }

    public T getAt(int index) {
        if(index > size){
            throw new RuntimeException("index out of range");
        }
        SNode<T> current = this.root;
        int _index = 0;
        while(current != null){
            if(_index == index) return current.value();
            current = current.next;
            _index++;
        }
        return null;
    }

    public void insertAt(int index,T data){
        if(index > size){
            throw new RuntimeException("index out of range");
        }
        if(index == size){
            append(data);
            return;
        }
        SNode<T> pre = null;
        SNode<T> current = this.root;
        int _index = 0;
        SNode<T> sNode = new SNode<T>(data);
        while(current != null){
            if(_index == index){
                sNode.next = current;
                if(pre == null){
                    this.root = sNode;
                }else{
                    pre.next = sNode;
                }
                this.size ++;
                return;
            }
            pre = current;
            current = current.next;
            _index ++;
        }
    }

    public T delete(T data) {
        SNode<T> pre = null;
        SNode<T> cur = this.root;
        while(cur != null){
            T value = cur.value();
            if(value.equals(data)){
                if(pre == null){
                    this.root = cur.next;
                }else{
                    pre.next = cur.next;
                }
                this.size --;
                return cur.value();
            }
        }
        
        return null;
    }

    public void show(){
        if(root == null)return;
        SNode<T> cNext = this.root.next;
        StringBuilder sBuilder = new StringBuilder();
        sBuilder.append("[");
        sBuilder.append(this.root.value());
        sBuilder.append("]");
        while(cNext != null){
            if(cNext != null) sBuilder.append("->");
            sBuilder.append("[");
            sBuilder.append(cNext.data);
            sBuilder.append("]");
            cNext = cNext.next;
        }
        System.out.println(sBuilder);
    }

    private static class SNode<T> {

        private T data;
        private SNode<T> next;

        private SNode(T data) {
            this.data = data;
        }

        public T value() {
            return data;
        }
    }

    public static void main(String[] args) {
        SingleLink<String> link = new SingleLink<>();
        System.out.println("---------append a b c------------");
        link.append("a").append("b").append("c");
        link.show();
        System.out.println("---------delete a b c d------------");
        link.delete("a");
        link.delete("b");
        link.delete("c");
        link.delete("d");
        System.out.println("size = "+link.size());
        link.show();
        System.out.println("---------insert at 0 = 10 & 0 =20 ------------");
        link.insertAt(0, "10");
        link.insertAt(0, "20");
        link.show();
        System.out.println("size = "+link.size());
        System.out.println("-----------for----------");
        for (int i = 0; i < link.size(); i++) {
            System.out.println(i + " = "+link.getAt(i));
        }
        System.out.println("-----------iterator----------");
        Iterator<String> it =  link.iterator();
        while(it.hasNext()){
            System.out.println("it next => " + it.next());
        }
        System.out.println("-----------again----------");
        it =  link.iterator();
        while(it.hasNext()){
            System.out.println("it next => " + it.next());
        }
        //it.next();
    }
}

File: lib/dlink/single_link.dart
// ignore_for_file: avoid_print

class SingleLink<T> with Iterator<T> {
  SNode<T>? _root;
  int _size = 0;

  SingleLink();

  SingleLink<T> append(T data) {
    SNode<T> node = SNode(data);
    if (_root == null) {
      _root = node;
    } else {
      SNode cNode = _root!;
      while (cNode.next != null) {
        cNode = cNode.next!;
      }
      cNode.next = node;
    }
    _size++;
    return this;
  }

  int get size => _size;
  SNode<T>? _currentNode;

  Iterator<T> iterator() {
    _currentNode = _root;
    return this;
  }

  @override
  T get current {
    if (_currentNode == null) {
      throw "out of range";
    }
    T value = _currentNode!.value()!;
    _currentNode = _currentNode!.next;
    return value;
  }

  @override
  bool moveNext() {
    return _currentNode != null;
  }

  T getAt(int index) {
    if (index > size) {
      throw "index out of range";
    }
    var current = _root;
    int _index = 0;
    while (current != null) {
      if (_index == index) return current.value();
      current = current.next;
      _index++;
    }
    throw "index out of range";
  }

  void insertAt(int index, T data) {
    if (index > size) {
      throw "index out of range";
    }
    if (index == size) {
      append(data);
      return;
    }
    SNode<T>? pre;
    var current = _root;
    int _index = 0;
    SNode<T> sNode = SNode<T>(data);
    while (current != null) {
      if (_index == index) {
        sNode.next = current;
        if (pre == null) {
          _root = sNode;
        } else {
          pre.next = sNode;
        }
        _size++;
        return;
      }
      pre = current;
      current = current.next;
      _index++;
    }
  }

  T delete(T data) {
    SNode<T>? pre;
    var cur = _root;
    while (cur != null) {
      T value = cur.value();
      if (value == data) {
        if (pre == null) {
          _root = cur.next;
        } else {
          pre.next = cur.next;
        }
        _size--;
        return cur.value();
      }
    }

    throw "not found";
  }

  void show() {
    if (_root == null) return;
    var cNext = _root!.next;
    StringBuffer sBuffer = StringBuffer();
    sBuffer.write("[");
    sBuffer.write(_root!.value());
    sBuffer.write("]");
    while (cNext != null) {
      sBuffer.write("->");
      sBuffer.write("[");
      sBuffer.write(cNext.data);
      sBuffer.write("]");
      cNext = cNext.next;
    }
    print(sBuffer);
  }
}

void main() {
  SingleLink<String> link = SingleLink<String>();
  print("---------append a b c------------");
  link.append("a").append("b").append("c");
  link.show();
  print("---------delete a b c------------");
  link.delete("a");
  link.delete("b");
  link.delete("c");
  //link.delete("d");
  print("size = ${link.size}");
  link.show();
  print("---------insert at 0 = 10 & 0 =20 ------------");
  link.insertAt(0, "10");
  link.insertAt(0, "20");
  link.show();
  print("size = ${link.size}");
  print("-----------for----------");
  for (int i = 0; i < link.size; i++) {
    print("$i = ${link.getAt(i)}");
  }
  print("-----------iterator----------");
  Iterator<String> it = link.iterator();
  while (it.moveNext()) {
    print("it next => ${it.current}");
  }
  print("-----------again----------");
  it = link.iterator();
  while (it.moveNext()) {
    print("it next => ${it.current}");
  }

  it.current;
}

class SNode<D> {
  D data;
  SNode<D>? next;

  SNode(this.data);

  D value() {
    return data;
  }
}
